src/languages/c-like.js

```diff
@@ -200,6 +201,58 @@ export default function(hljs) {
     ]
   };
 
+  var TEMPLATE_USE = {
+    begin: /</, end: />/,
+    keywords: CPP_KEYWORDS,
+    contains: EXPRESSION_CONTAINS.concat([
+      // Match left-shift to prevent it from creating a nested `TEMPLATE_USE`.
+      { begin: /<</ },
+      'self'
+    ])
+  };
+
+  var CLASS_DECLARATION = {
+    className: 'class',
+    beginKeywords: 'class struct union', end: /[{;:]/,
+    contains: [
+      { beginKeywords: "final" },
+      TEMPLATE_USE,
+      TITLE_MODE
+    ]
+  };
+
+  var TEMPLATE_DECLARATION = {
+    beginKeywords: 'template',
+    // Add additional stops to stop runaway template declaration, e.g.
+    // `template <bool b = 1 < 2> void f();`
+    end: /[>;{]/,
+    contains: [
+      {
+        begin: /</, end: />/,
+        endsWithParent: true,
```

Comment:
Just so you know this `endsWithParent` does not nest... it simply means that within the scope of a `<...>` that it will attempt to match not on the contains rules on line 235, but ALSO the end rule on line 228.  But it doesn't go any deeper than that unless you added `endsWithParent` at every level...

I'm not sure if this has implications for your runaway guard on line 227 or not...

Reply:
Good question! So, if I understand you correctly, I could equivalently remove `endsWithParent, and change the `end` rule to `/[>;{]/` (i.e. the same as the parent `end`)? In this case I should probably do that instead, and add the same end to the other contained rules.

Reply:
Well if the desire is that they end with the parent you'd probably want to just use that... that's what it does - pull the parent "end" down into the children... I was just pointing out that it only goes one layer deep... so if it got "stuck" in a child of a child then your "guard" would not save you... so that needs to be considered and might result in adding some more "false positive" edge case tests.

Reply:
I think that works now. I needed to add a nested `{begin: />/}` to prevent a `>` inside a `(…)` from terminating the template declaration mode prematurely. That said, the behaviour on my unit tests is actually the same as before, I need to think of some example which works now and would previously have failed.

Reply:
Damn, I think I’ve just kicked the can down the road now. I think this fundamentally doesn’t work because I would need to nest the parent mode inside the mode that handles `(…)`. With my newest change, the the two first lines parse correctly, but the template declaration in the third line ends prematurely, and the “class title” starts at `class T` rather than at `class C`:

```cpp
template <int A = (B<C>), class T> class C;
template <int A = B<(c > D<E>)>, class T> class C;
template <int A = B<(c > D<(e > f)>)>, class T> class C;
```

Reply:
```
  var TEMPLATE_USE = {};
  Object.assign(TEMPLATE_USE, {
    // .. can now use TEMPLATE_USE recursively
```

All you need is a reference. :-)  Though I'm not sure we do this anywhere else and not sure it's a pattern we should start using now though. ...

I'm guessing the nesting here could be unlimited but is there some practical limits for MOST C++ code as far as what you'd see in the real world?